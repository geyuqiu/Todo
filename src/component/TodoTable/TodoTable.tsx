import React, {useState} from "react";

import {tableColumns} from '../Row/TableColumns';
import {Row} from '../Row/Row';
import {Table} from '../Table/Table';
import {Todo} from '../Row/model';
import {Filters} from '../Filters/Filters';

type TableProps = {
	todos: Todo[];
	isLoading: boolean;
}

export const TodoTable = ({todos, isLoading}: TableProps) => {
	const [title, setTitle] = useState<string>("");
	const [completed, setCompleted] = useState<string>("");
	return (
		<div className="ml-6 mr-3 sm:mx-8 pt-12">
			<h1 className="font-black mb-4 text-3xl">Todos</h1>
			<Filters title={title} completed={completed} setTitle={setTitle} setCompleted={setCompleted}/>
			{(isLoading) && <p data-testid="Table__Loading">Loading ...</p>}
			{!isLoading && !todos?.length &&
			<NoResults/>
			}
			{!isLoading && todos?.length > 0 &&
			<Table columns={tableColumns} data={todos}
						 titleOutside={title} completedOutside={completed}
			>
				{(todo: Todo) => (
					<Row todo={todo}/>
				)}
			</Table>
			}
		</div>
	);
}

export const NoResults = () =>
	<p data-testid="Table__No_Results" className="bg-error p-4 text-white">No result for the given
	search parameters!</p>
