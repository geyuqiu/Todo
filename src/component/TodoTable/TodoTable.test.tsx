import {render} from "@testing-library/react";
import React from "react";

import {NoResults, TodoTable} from "./TodoTable";

export const todos = [
	{
		id: 1,
		title: "title 1",
		completed: 'yes'
	},
	{
		id: 2,
		title: "title 2",
		completed: 'yes'
	},
	{
		id: 3,
		title: "title 3",
		completed: 'yes'
	},
	{
		id: 4,
		title: "delectus aut autem",
		completed: 'no'
	}
];

describe("TodoTable", () => {
	it("NoResults should match snapshot", () => {
		const {getAllByTestId, asFragment} = render(<NoResults/>);
		expect(getAllByTestId("Table__No_Results")).toHaveLength(1);
		expect(asFragment()).toMatchSnapshot();
	});

	it("isLoading should match snapshot", () => {
		const {getAllByTestId, asFragment} = render(<TodoTable isLoading/>);
		expect(getAllByTestId("Table__Loading")).toHaveLength(1);
		expect(asFragment()).toMatchSnapshot();
	});

	it("todos empty for selected address should match snapshot", () => {
		const {getAllByTestId, asFragment} = render(<TodoTable
			todos={[]} isLoading={false}/>);

		expect(getAllByTestId("Table__No_Results")).toHaveLength(1);
		expect(asFragment()).toMatchSnapshot();
	});

	it("display n todos should match snapshot", () => {
		const {getAllByTestId, asFragment} = render(<TodoTable
			todos={todos} isLoading={false}
		/>);

		expect(getAllByTestId("TableRow")).toHaveLength(4);
		expect(asFragment()).toMatchSnapshot();
	});
});
