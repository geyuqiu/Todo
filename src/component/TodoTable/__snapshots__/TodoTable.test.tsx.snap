// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoTable NoResults should match snapshot 1`] = `
<DocumentFragment>
  <p
    class="bg-error p-4"
    data-testid="Table__No_Results"
  >
    No result for the given search parameters!
  </p>
</DocumentFragment>
`;

exports[`TodoTable display n todos should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="ml-6 mr-3 sm:mx-8 pt-12"
  >
    <div
      class="Table__TableWrapper-xscllg-0 gqOIbh w-full max-w-screen-xl"
      role="table"
    >
      <table
        cellpadding="0"
        class="table-auto"
      >
        <thead>
          <tr
            class="border"
            role="row"
          >
            <th
              class="group border border-gray-lightest bg-gray-dark text-base text-center select-none m-0 p-3 md:px-5 font-semibold"
              colspan="1"
              data-testid="table__th--0"
              role="columnheader"
            >
              #
            </th>
            <th
              class="group border border-gray-lightest bg-gray-dark text-base text-center select-none m-0 p-3 md:px-5 font-semibold"
              colspan="1"
              data-testid="table__th--1"
              role="columnheader"
            >
              Title
            </th>
            <th
              class="group border border-gray-lightest bg-gray-dark text-base text-center select-none m-0 p-3 md:px-5 font-semibold"
              colspan="1"
              data-testid="table__th--2"
              role="columnheader"
            >
              Completed
            </th>
          </tr>
        </thead>
        <tbody
          role="rowgroup"
        >
          <tr
            class="TableRow__TableRowStyled-sc-990v33-0 iJGVRV"
            data-testid="TableRow"
          >
            <td
              class="border-gray-lightest border"
            >
              <div
                class="TableCell__TableCellInnerWrapper-l6uokn-0 dwniYi"
              >
                1
              </div>
            </td>
            <td
              class="border-gray-lightest border"
            >
              <div
                class="TableCell__TableCellInnerWrapper-l6uokn-0 dwniYi"
              >
                title 1
              </div>
            </td>
            <td
              class="border-gray-lightest border"
            >
              <div
                class="TableCell__TableCellInnerWrapper-l6uokn-0 dwniYi"
              >
                true
              </div>
            </td>
          </tr>
          <tr
            class="TableRow__TableRowStyled-sc-990v33-0 iJGVRV"
            data-testid="TableRow"
          >
            <td
              class="border-gray-lightest border"
            >
              <div
                class="TableCell__TableCellInnerWrapper-l6uokn-0 dwniYi"
              >
                2
              </div>
            </td>
            <td
              class="border-gray-lightest border"
            >
              <div
                class="TableCell__TableCellInnerWrapper-l6uokn-0 dwniYi"
              >
                title 2
              </div>
            </td>
            <td
              class="border-gray-lightest border"
            >
              <div
                class="TableCell__TableCellInnerWrapper-l6uokn-0 dwniYi"
              >
                false
              </div>
            </td>
          </tr>
          <tr
            class="TableRow__TableRowStyled-sc-990v33-0 iJGVRV"
            data-testid="TableRow"
          >
            <td
              class="border-gray-lightest border"
            >
              <div
                class="TableCell__TableCellInnerWrapper-l6uokn-0 dwniYi"
              >
                3
              </div>
            </td>
            <td
              class="border-gray-lightest border"
            >
              <div
                class="TableCell__TableCellInnerWrapper-l6uokn-0 dwniYi"
              >
                title 3
              </div>
            </td>
            <td
              class="border-gray-lightest border"
            >
              <div
                class="TableCell__TableCellInnerWrapper-l6uokn-0 dwniYi"
              >
                true
              </div>
            </td>
          </tr>
          <tr
            class="TableRow__TableRowStyled-sc-990v33-0 iJGVRV"
            data-testid="TableRow"
          >
            <td
              class="border-gray-lightest border"
            >
              <div
                class="TableCell__TableCellInnerWrapper-l6uokn-0 dwniYi"
              >
                4
              </div>
            </td>
            <td
              class="border-gray-lightest border"
            >
              <div
                class="TableCell__TableCellInnerWrapper-l6uokn-0 dwniYi"
              >
                delectus aut autem
              </div>
            </td>
            <td
              class="border-gray-lightest border"
            >
              <div
                class="TableCell__TableCellInnerWrapper-l6uokn-0 dwniYi"
              >
                false
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TodoTable isLoading should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="ml-6 mr-3 sm:mx-8 pt-12"
  >
    <p
      data-testid="Table__Loading"
    >
      Loading ...
    </p>
  </div>
</DocumentFragment>
`;

exports[`TodoTable todos empty for selected address should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="ml-6 mr-3 sm:mx-8 pt-12"
  >
    <p
      class="bg-error p-4"
      data-testid="Table__No_Results"
    >
      No result for the given search parameters!
    </p>
  </div>
</DocumentFragment>
`;
